blueflood-core/src/main/java/com/rackspacecloud/blueflood/inputs/processors/TokenDiscoveryWriter.java

```diff
@@ -0,0 +1,133 @@
+package com.rackspacecloud.blueflood.inputs.processors;
+
+import com.codahale.metrics.Meter;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.rackspacecloud.blueflood.cache.LocatorCache;
+import com.rackspacecloud.blueflood.cache.TokenCache;
+import com.rackspacecloud.blueflood.concurrent.FunctionWithThreadPool;
+import com.rackspacecloud.blueflood.io.TokenDiscoveryIO;
+import com.rackspacecloud.blueflood.service.Configuration;
+import com.rackspacecloud.blueflood.service.CoreConfig;
+import com.rackspacecloud.blueflood.types.IMetric;
+import com.rackspacecloud.blueflood.types.Locator;
+import com.rackspacecloud.blueflood.types.Token;
+import com.rackspacecloud.blueflood.utils.Metrics;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * A separate discovery writer for tokens. This class is modelled after {@link DiscoveryWriter}
+ */
+public class TokenDiscoveryWriter extends FunctionWithThreadPool<List<List<IMetric>>, Void> {
+
+    private final List<TokenDiscoveryIO> tokenDiscoveryIOs = new ArrayList<>();
```

Comment:
This is the extensibility that blueflood provides which was existing and I merely supported it. We can provide any implementations for `TokenDiscoveryIO`. I only provided one implementation `ElasticTokensIO` which is powered by elastic search. But if you want to provided a different implementation say for graph database, all you have to do is provide another implementation for `TokenDiscoveryIO` using graphs and we are good.